File: 216. Combination Sum III.cpp
class Solution {
public:
    void backtrack(vector<vector<int>>& ans, unordered_set<string>& seen,
                   int currEle, vector<int>& currArr, int currSum,
                   int sum, int count, int maxCount) {

        if (currSum > sum || count > maxCount) return;
        if (currEle > 9 || (currSum == sum && count == maxCount)) {
            if (currSum == sum && count == maxCount) {
                string key;
                for (int num : currArr) {
                    key += to_string(num) + ",";
                }
                if (seen.count(key)) return;
                seen.insert(key);
                ans.push_back(currArr);
            }
            return;
        }

        currArr.push_back(currEle);
        backtrack(ans, seen, currEle + 1, currArr, currSum + currEle, sum, count + 1, maxCount);
        currArr.pop_back();

        backtrack(ans, seen, currEle + 1, currArr, currSum, sum, count, maxCount);
    }

    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        unordered_set<string> seen;
        vector<int> currArr;

        backtrack(ans, seen, 1, currArr, 0, n, 0, k);
        return ans;
    }
};

File: 216. Combination Sum III.java
class Solution {
    public void backtrack(List<List<Integer>> ans, HashSet<ArrayList<Integer>> hs, int currEle, ArrayList<Integer> currArr, int currSum, int sum, int count, int maxCount){
        if(currSum > sum || count > maxCount) return;
        if(currEle > 9 || (currSum == sum && count == maxCount)){
            if(currSum == sum && count == maxCount){
                if(hs.contains(currArr)) return;
                hs.add(new ArrayList<>(currArr));
                ans.add(new ArrayList<>(currArr));
            }
            return;
        }
        currArr.add(currEle);
        backtrack(ans, hs, currEle + 1, currArr, currSum + currEle, sum, count + 1, maxCount);

        currArr.remove(currArr.size() - 1);
        backtrack(ans, hs, currEle + 1, currArr, currSum, sum, count, maxCount);
    }
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> ans = new ArrayList<>();
        HashSet<ArrayList<Integer>> hs = new HashSet<>();

        backtrack(ans, hs, 1, new ArrayList<>(),0, n, 0, k);
        
        return ans;
    }
}